block-storage/src/main/scala/coop/rchain/blockstorage/FileLMDBIndexBlockStore.scala

```diff
@@ -0,0 +1,255 @@
+                 }
+      } yield result
+    )
+
+  override def put(f: => (BlockHash, BlockMessage)): F[Unit] =
+    lock.withPermit(
+      for {
+        blockMessageRandomAccessFile <- blockMessageRandomAccessFileRef.get
+        endOfFileOffset              <- Sync[F].delay { blockMessageRandomAccessFile.length() }
+        _                            <- Sync[F].delay { blockMessageRandomAccessFile.seek(endOfFileOffset) }
+        (blockHash, blockMessage)    = f
+        blockMessageByteArray        = blockMessage.toByteArray
+        _                            <- Sync[F].delay { blockMessageRandomAccessFile.writeInt(blockMessageByteArray.length) }
+        _                            <- Sync[F].delay { blockMessageRandomAccessFile.write(blockMessageByteArray) }
+        _ <- withWriteTxn { txn =>
+              index.put(
+                txn,
+                blockHash.toDirectByteBuffer,
+                endOfFileOffset.toByteString.toDirectByteBuffer
+              )
+            }
+      } yield ()
+    )
+
+  override def checkpoint(): F[Unit] =
+    ().pure[F]
+
+  override def clear(): F[Unit] =
+    lock.withPermit(
+      for {
+        blockMessageRandomAccessFile <- blockMessageRandomAccessFileRef.get
+        _                            <- Sync[F].delay { blockMessageRandomAccessFile.setLength(0) }
+        _ <- withWriteTxn { txn =>
+              index.drop(txn)
+            }
+      } yield ()
+    )
+
+  override def close(): F[Unit] =
+    lock.withPermit(
+      for {
+        blockMessageRandomAccessFile <- blockMessageRandomAccessFileRef.get
+        _                            <- Sync[F].delay { blockMessageRandomAccessFile.close() }
+      } yield ()
+    )
+}
+
+object FileLMDBIndexBlockStore {
+  private val checkpointPattern: Regex = "([0-9]+)-([0-9]+)".r
+
+  case class Config(
+      storagePath: Path,
+      indexPath: Path,
+      checkpointsDirPath: Path,
+      mapSize: Long,
+      maxDbs: Int = 1,
+      maxReaders: Int = 126,
+      noTls: Boolean = true
+  )
+
+  private[blockstorage] case class CheckpointIndex(
+      env: Env[ByteBuffer],
+      index: Dbi[ByteBuffer]
+  )
+
+  private[blockstorage] case class Checkpoint(
+      start: Long,
+      end: Long,
+      dirPath: Path,
+      storagePath: Path,
+      indexPath: Path,
+      index: Option[WeakReference[CheckpointIndex]]
+  )
+
+  private def loadCheckpoints[F[_]: Sync: Log](checkpointsDirPath: Path): F[List[Checkpoint]] =
+    for {
+      checkpointDirectories <- Sync[F].delay {
+                                checkpointsDirPath.toFile.mkdir()
+                                Files.list(checkpointsDirPath).filter(p => Files.isDirectory(p))
+                              }
+      checkpointDirectoriesList = checkpointDirectories
+        .collect(Collectors.toList[Path])
+        .asScala
+        .toList
+      checkpoints <- checkpointDirectoriesList.flatTraverse { dirPath =>
```

Comment:
you should use just `traverse` (instead of `flatTraverse`) here are all the elements that your function is returning are single elements (which you artificially wrap into `List`

Reply:
One of the branches returns an empty `List`. The semantics here is that some of the files inside of the checkpoints directory could be named differently from what is expected and should be ignored. If the resulting list of checkpoint files is still satisfactory (i.e. starts from 0 and is consecutive) then we can proceed forward. This also answers you question about stopping rnode. I don't think it is a fatal error as having a *consistent* list of checkpoints is a necessary condition, not having a pure checkpoints directory.